Add support for MPRIS2 DBus spec

RobertMe opened the pull request:
Fixes #37 

Implemented MPRIS2 support in the library, so that it can be used by an app using just a couple of lines of code.

- [x] Add basic API for `MediaPlayer2` and `MediaPlayer2.Player` interfaces
- [x] Send basic metadata
- [x] Support seeking etc
- [x] Hook up in Sailfish
- [x] Implement support for all media types (currently episode only)
  - [x] Add/handle paths for other types
  - [x] Expand meta data with details of other types
- [x] Implement minor missing options/properties (`LoopStatus`, `Shuffle`, `Volume`)
- [x] Implement OpenUrl support (at least basic support for some kind of `youtube://<video id>`, playback of local items would be too much work, as it needs something to stream/share the item to Kodi)

Use cases (for Sailfish) are (hopefully as all are untested):
- Playback control on lock screen, when feature is released by Jolla and hopefully uses MPRIS2
- Playback control using the Pebble watch (for example using https://github.com/smokku/pebble)
- Allowing other applications to start playback on Kodi (like [Ytplayer](https://github.com/tworaz/sailfish-ytplayer))

RobertMe commented:
Note to self: move `MprisPlayer::OpenUrl` implementation to `ProtocolManager(::execute?)`. As it will (/must) be usable in other areas, like a command line option.

mzanetti commented:
Fails to compile. The ctor MprisController doesn't allow it to be called as you do. Fixing that, there are still linker errors. Probably QT += dbus missing in the lib?

mzanetti commented on libkodimote/libkodimote.pro:
```diff
@@ -75,7 +76,14 @@ SOURCES +=  kodi.cpp \
             profiles.cpp \
-            profileitem.cpp
+            profileitem.cpp \
+    mpris2/mpriscontroller.cpp \
```
indentation

RobertMe commented:
> The ctor MprisController doesn't allow it to be called as you do.

Right, I tested in a dummy app which passed in the param, but didn't check Sailfish recently

> Probably QT += dbus missing in the lib?

Right too, would be the `Qt` vs `QT` as stated in #61

Both should be fixed

mzanetti commented on libkodimote/mpris2/mprisplayer.h:
```diff
@@ -0,0 +1,90 @@
+    void Stop();
+    void Seek(qint64 offset);
+    void SetPosition(QDBusObjectPath path, qint64 position);
+    void OpenUrl(QString uri);
```
should be openUri(), not openUrl() afaict.

mzanetti commented on libkodimote/mpris2/mprisapplication.h:
```diff
@@ -0,0 +1,43 @@
+    Q_OBJECT
+    Q_PROPERTY(bool CanQuit READ canQuit)
+    Q_PROPERTY(bool CanSetFullscreen READ canSetFullscreen)
+    Q_PROPERTY(bool canRaise READ canRaise)
```
CanRaise

mzanetti commented:
So I've been playing around for a while with this but still haven't managed to make it show up in Ubuntu's sound indicator. Everything else implementing MPRIS works fine with that though. Because of that, and some typos in the API I'm starting to think it would make sense to include this file in the repo and use qdbusxml2cpp on it to generate the api interface: http://cgit.freedesktop.org/xdg/mpris-spec/tree/spec/org.mpris.MediaPlayer2.Player.xml

Also would make updating it to a newer mpris version easier and more predictible if that should happen.

mzanetti commented on libkodimote/mpris2/mprisapplication.h:
```diff
@@ -0,0 +1,43 @@
+    QStringList supportedMimeTypes() const;
+
+    bool fullscreen() const;
+    void setFullscreen(bool fullscreen);
```
Q_PROPERTY for Fullscreen is missing

mzanetti commented:
Ok. have it working in Ubuntu's indicators now. Need some further testing. Have some issues with the play status not updating, but not sure yet if that's related to the MPRIS stuff or perhaps an issue we have somewhere else.

Here's a screenshot: http://i.imgur.com/r1VpogS.png

Bonus points if you add the nowplaying stuff like rhytmbox does ;)

mzanetti commented:
Here's another screenshot of the MPRIS playlist interface: http://i.imgur.com/690Z3nf.png

Would be nice to have that too, at least for recently added and recently played I guess we'd just need to wire up existing stuff. The other two might come in nicely with your other branch?

RobertMe commented:
Rebased again so it's a bit cleaner, fixed the above issues also.

Don't know about Ubuntu not showing the metadata.

I've also added the `mpris:artUrl` metadata, so it shows the thumbnail.
At least at the moment I'm not that interested in implementing the `TrackList` and (requested) `Playlists` interfaces, because I don't have any use cases, nor see them, myself (which would also make testing harder). But feel free to add them.

mzanetti commented on libkodimote/mpris2/mprisplayer.cpp:
```diff
@@ -0,0 +1,394 @@
+        return QString("/org/mpris/MediaPlayer2/Track/episode/%1").arg(item->episodeId());
+    } else if (item->movieId() != -1) {
+        return QString("/org/mpris/MediaPlayer2/Track/movie/%1").arg(item->movieId());
+    } else if (item->musicvideoId()) {
```
!= -1 missing.

mzanetti commented on libkodimote/mpris2/mprisplayer.cpp:
```diff
@@ -0,0 +1,394 @@
+    QVariantMap map;
+    if (m_player && m_player->currentItem()) {
+        LibraryItem *item = m_player->currentItem();
+        map["mpris:trackid"] = qVariantFromValue(QDBusObjectPath(buildPath(item)));
```
This breaks the metadata call and is the reason why metadata doesn't show up in ubuntu. Removing the qVariantFromValue(QDBusObjectPath()) seems to fix it.

mzanetti commented on libkodimote/mpris2/mprisplayer.cpp:
```diff
@@ -0,0 +1,394 @@
+        map["mpris:trackid"] = qVariantFromValue(QDBusObjectPath(buildPath(item)));
+        map["mpris:length"] = QTime(0, 0, 0).msecsTo(item->duration()) * Q_INT64_C(1000); // milliseconds vs microseconds
+        map["xesam:title"] = item->title();
+        map["xesam:artist"] = item->subtitle();
```
item->subtitle() doesn't seems to work. Looking that the NowPlayingPage this needs to check the type and explicitly fetch artist, video ratings etc.

RobertMe replied:
AFAICT this should work, as subtitle is being set in `Player` ( https://github.com/mzanetti/kodimote/blob/212cc0451f10fa08424786d50e060a3e1edcd114/libkodimote/player.cpp#L416 ), and it's working for me on episodes of TV Shows (not recordings :stuck_out_tongue_closed_eyes: ). So could you check if you aren't playing an item which doesn't have a subtitle set?

mzanetti commented on libkodimote/mpris2/mprisplayer.cpp:
```diff
@@ -0,0 +1,394 @@
+            disconnect(item, SIGNAL(thumbnailChanged()), this, SLOT(thumbnailChanged()));
+            connect(item, SIGNAL(thumbnailChanged()), this, SLOT(thumbnailChanged()));
+        } else if (!thumbnail.isEmpty()) {
+            map["mpris:artUrl"] = item->thumbnail();
```
requires a url, in this case, prefixing it with "file://"

RobertMe commented:
Will add Ubuntu support tomorrow, and also update the `(Can)Raise` behavior as discussed on IRC then.

RobertMe commented:
Added missing license text to the top of the files and added raise support based on platform helpers.

Only did some quick testing, so proper testing is still needed.

mzanetti commented on libkodimote/mpris2/mprisplayer.cpp:
```diff
@@ -0,0 +1,420 @@
+        disconnect(m_player, SIGNAL(speedChanged()), this, SLOT(speedChanged()));
+        disconnect(m_player, SIGNAL(shuffleChanged()), this, SLOT(shuffleChanged()));
+        disconnect(m_player, SIGNAL(repeatChanged()), this, SLOT(repeatChanged()));
+        disconnect(m_player->playlist(), SIGNAL(countChanged()), this, SLOT(playlistChanged()));
```
I think you can just use this one: disconnect(myObject, 0, myReceiver, 0); like:
disconnect(m_player, 0, this, 0); then you don't need to list them all.

RobertMe replied:
Didn't know about that, changed.

mzanetti commented on libkodimote/mpris2/mprisplayer.cpp:
```diff
@@ -0,0 +1,420 @@
+
+void MprisPlayer::currentItemChanged()
+{
+    connect(m_player->currentItem(), SIGNAL(thumbnailChanged()), this, SLOT(thumbnailChanged()));
```
This crashes when disconnecting from kodi. Here's a trace: http://paste.ubuntu.com/10576993/

is there a reason why you connect this one in here and not in line 370, inside the if(m_player) like all the others?

RobertMe replied:
This should have been `disconnect`, so it disconnects when you start playing another item. But `player->currentItem()` would be invalid at that point, so I added a new member to track the currentItem the connection was set to (although that might not be needed, as the item gets destroyed inside Player).
